nntrainer/tensor/blas_neon.cpp

```diff
@@ -0,0 +1,267 @@
+
+      const float *__restrict w;
+
+      float32x4_t y0;
+
+      for (unsigned int j = 0; j < rows; ++j) {
+        w = &A[j * cols + i];
+        y0 = vmovq_n_f32(0);
+
+        float r[4];
+        wvec0_3 = vld1q_f32(&w[0]);
+        wvec4_7 = vld1q_f32(&w[4]);
+
+        y0 = vmlaq_f32(y0, wvec0_3, x0_3);
+        y0 = vmlaq_f32(y0, wvec4_7, x4_7);
+
+        vst1q_f32(r, y0);
+        for (unsigned int k = 0; k < 4; ++k) {
+          Y[j] = Y[j] + r[k];
+        }
+      }
+    }
+  } else if (cols % 4 == 0) {
+    for (unsigned i = 0; i < cols; i += 4) {
+      float32x4_t x0_3 = vld1q_f32(&X[i]);
+
+      if (alpha != 1.0) {
+        x0_3 = vmulq_f32(x0_3, v_alpha);
+      }
+
+      float32x4_t wvec0_3, wvec4_7;
+
+      const float *__restrict w;
+
+      float32x4_t y0;
+
+      for (unsigned int j = 0; j < rows; ++j) {
+        w = &A[j * cols + i];
+        y0 = vmovq_n_f32(0);
+
+        float r[4];
+        wvec0_3 = vld1q_f32(&w[0]);
+
+        y0 = vmlaq_f32(y0, wvec0_3, x0_3);
+
+        vst1q_f32(r, y0);
+        for (unsigned int k = 0; k < 4; ++k) {
+          Y[j] = Y[j] + r[k];
+        }
+      }
+    }
+  }
+}
+
+void sgemv_transpose_neon(const float *A, const float *X, float *Y,
+                          uint32_t rows, uint32_t cols, float alpha,
+                          float beta) {
+  const float *__restrict x;
+
+  const float32x4_t v_beta = vdupq_n_f32(beta);
+  const float32x4_t v_alpha = vdupq_n_f32(alpha);
+
+  if (cols % 16 == 0) {
+    bool initialized[cols / 16];
+    unsigned int step;
+    for (unsigned int i = 0; i < cols / 16; ++i) {
+      initialized[i] = false;
+    }
+
+    for (unsigned int i = 0; i < rows; ++i) {
+      float32x4_t x = vld1q_dup_f32(&X[i]);
+      x = vmulq_f32(x, v_alpha);
+
+      for (unsigned int j = 0; j < cols; j += 16) {
+        float *__restrict y = &Y[j];
+
+        float32x4_t y0_3 = vld1q_f32(&y[0]);
+        float32x4_t y4_7 = vld1q_f32(&y[4]);
+        float32x4_t y8_11 = vld1q_f32(&y[8]);
+        float32x4_t y12_15 = vld1q_f32(&y[12]);
+        step = j / 16;
+        if (!initialized[step]) {
+          y0_3 = vmulq_f32(y0_3, v_beta);
+          y4_7 = vmulq_f32(y4_7, v_beta);
+          y8_11 = vmulq_f32(y8_11, v_beta);
+          y12_15 = vmulq_f32(y12_15, v_beta);
+          initialized[step] = true;
+        }
+
+        float32x4_t wvec0_3, wvec4_7, wvec8_11, wvec12_15;
+        const float *__restrict w;
+
+        w = &A[i * cols + j];
+
+        wvec0_3 = vld1q_f32(&w[0]);
+        wvec4_7 = vld1q_f32(&w[4]);
+        wvec8_11 = vld1q_f32(&w[8]);
+        wvec12_15 = vld1q_f32(&w[12]);
+
+        y0_3 = vmlaq_f32(y0_3, wvec0_3, x);
+        y4_7 = vmlaq_f32(y4_7, wvec4_7, x);
+        y8_11 = vmlaq_f32(y8_11, wvec8_11, x);
+        y12_15 = vmlaq_f32(y12_15, wvec12_15, x);
+
+        vst1q_f32(&y[0], y0_3);
+        vst1q_f32(&y[4], y4_7);
+        vst1q_f32(&y[8], y8_11);
+        vst1q_f32(&y[12], y12_15);
+      }
+    }
+    return;
```

Comment:
It seems to be unnecessary for `return` for all if else statement

Reply:
I think it needs. %16 == 0 also true for %8 and %4.

Reply:
I mean it can return naturally in line 264, and 'return' in line 264 seems unnecessary, too.